integration/test/metrics_number_dimension/metrics_number_dimension_test.go

```diff
@@ -0,0 +1,133 @@
+)
+
+const configOutputPath = "/opt/aws/amazon-cloudwatch-agent/bin/config.json"
+const configJSON = "/config.json"
+const namespace = "MetricNumberDimensionTest"
+const instanceId = "InstanceId"
+const append = "append"
+
+//Let the agent run for 2 minutes. This will give agent enough time to call server
+const agentRuntime = 2 * time.Minute
+
+var targetString = "max MaxDimensions %v is less than than number of dimensions %v thus only taking the max number"
+
+type input struct {
+	dataInput            string
+	findTarget           bool
+	inputNumberDimension int
+	metricName           string
+}
+
+type metric struct {
+	name  string
+	value string
+}
+
+//Must run this test with parallel 1 since this will fail if more than one test is running at the same time
+//This test uses a pem file created for the local stack endpoint to be able to connect via ssl
+func TestBundle(t *testing.T) {
+
+	parameters := []input{
+		{dataInput: "resources/10_dimension",
+			findTarget:           false,
+			inputNumberDimension: 10,
+			metricName:           "mem_used_percent"},
+		// @TODO add when https://github.com/aws/amazon-cloudwatch-agent/pull/361 is merged
+		// {dataInput: "resources/30_dimension", findTarget: false, inputNumberDimension: 30, metricName: "mem_used_percent"},
+		// {dataInput: "resources/35_dimension", findTarget: true, inputNumberDimension: 35, metricName: "mem_used_percent"},
+	}
+
+	for _, parameter := range parameters {
+		//before test run
+		log.Printf("resource file location %s find target %t input number dimension %d metric name %s",
+			parameter.dataInput, parameter.findTarget, parameter.inputNumberDimension, parameter.metricName)
+
+		targetString = fmt.Sprintf(targetString, 30, parameter.inputNumberDimension)
+
+		t.Run(fmt.Sprintf("resource file location %s find target %t", parameter.dataInput, parameter.findTarget), func(t *testing.T) {
+			test.CopyFile(parameter.dataInput+configJSON, configOutputPath)
+			test.StartAgent(configOutputPath)
+			time.Sleep(agentRuntime)
+			log.Printf("Agent has been running for : %s", agentRuntime.String())
+			test.StopAgent()
+
+			// test for target string
+			output := test.ReadAgentOutput(agentRuntime)
+			containsTarget := outputLogContainsTarget(output)
+			if (parameter.findTarget && !containsTarget) || (!parameter.findTarget && containsTarget) {
+				t.Errorf("Find target is %t contains target is %t", parameter.findTarget, containsTarget)
+			}
+
+			// test for cloud watch metrics
+			cxt := context.Background()
+			dimensionFilter := buildDimensionFilterList(parameter.inputNumberDimension - 1)
+			client := test.GetCWClient(cxt)
+			listMetricsInput := cloudwatch.ListMetricsInput{
+				MetricName: aws.String(parameter.metricName),
+				Namespace:  aws.String(namespace),
+				Dimensions: dimensionFilter,
+			}
+			data, err := client.ListMetrics(cxt, &listMetricsInput)
+			if err != nil {
+				t.Errorf("Error getting metric data %v", err)
+			}
+			if len(data.Metrics) == 0 {
```

Comment:
Is there any way to get a consistent/semi-consistent estimate of the expected number of metrics? Basically this is the same as saying when we ListMetrics, we expect at least one metric to come back, which doesn't tell us much - hypothetical situation, but assuming that we had a bug in the metric publishing when we try to append more than 30 dimensions that causes the agent to die after submitting the metrics once, this would not help us catch that (maybe it shouldn't?). 

I'd prefer this to be fleshed out more as a generalized validation function, but maybe I'm just biased because that's how I did the log validation.

Reply:
We would catch this by the finding the debug line in the log.

Reply:
Also agent does not fail on more than 30 metrics it just strips dimension to 30 then continues